### PR TITLE
feat: Add multi-backend support (CUDA/IX/MUSA/NPU) and expand C++ ope…

### PR DESCRIPTION
**Pull request by chen98038**
### Title
  Add multi-backend abstraction layer and expand C++ operator bindings

  ### PR Category
  Refactor

  ### Type of Change
  New Feature | Refactor

  ### Description

  This PR introduces a backend abstraction layer to decouple all C++ operator implementations
  from hardcoded CUDA dependencies, enabling support for multiple hardware backends (CUDA, IX,
  NPU/Ascend, MUSA/Moore Threads).

  **Key changes:**

  1. **Backend abstraction (`backend_utils.h`)**
     - New header `include/flag_gems/backend_utils.h` providing `backend::StreamType`,
  `backend::RawStreamType`, `backend::getCurrentStream()`, `backend::getRawStream()`,
  `backend::getDefaultDevice()`, etc.
     - Replaces direct `c10::cuda::CUDAStream` / `CUstream` usage across all 20+ operator files
  in `lib/`.

  2. **CMake multi-backend build system**
     - New cache variable `FLAGGEMS_BACKEND` (CUDA | IX | MUSA | NPU) controls backend
  selection.
     - Conditional `project()` language setup (skip CUDA language for NPU/MUSA).
     - Per-backend compile definitions (`FLAGGEMS_USE_CUDA`, `FLAGGEMS_USE_IX`,
  `FLAGGEMS_USE_NPU`, `FLAGGEMS_USE_MUSA`).
     - New `cmake/BackendNPU.cmake` and `cmake/BackendMUSA.cmake` modules for Ascend and Moore
  Threads toolchain discovery.
     - Backend-conditional linking in `lib/CMakeLists.txt`.

  3. **Dispatch key abstraction**
     - `TORCH_LIBRARY_IMPL` now uses `FLAGGEMS_DISPATCH_KEY` macro (CUDA for CUDA/IX,
  PrivateUse1 for NPU/MUSA) instead of hardcoded `CUDA`.

  4. **Expanded pybind11 bindings (`cstub.cpp`)**
     - Newly exposed operators: `exponential_`, `addmm`, `mm`, `zeros`, `topk`, `contiguous`,
  `cat`, `bmm`, `embedding`, `embedding_backward`, `argmax`, `sort`, `sort_stable`, `softmax`,
  `softmax_backward`, `reshape_and_cache_flash`, `flash_attn_varlen_func`.
     - Added lambda wrappers for `sum_dim` (OptionalIntArrayRef conversion) and `cat` /
  `contiguous` (default argument handling).

  5. **NPU (Ascend) specific adaptations**
     - NPU-specific Triton kernels: `mm_cpp.py`, `nonzero.py`, `rms_norm.py`,
  `fused_add_rms_norm.py`.
     - `float` casts for `epsilon` parameters (BiSheng MLIR does not support fp64).
     - Smaller block sizes for NPU UB memory constraints (e.g., MM uses 32x32x32).
     - Disabled StreamK path for NPU in `mm.cpp`.
     - `torch.utils.rename_privateuse1_backend("npu")` integration.

  6. **C++ standard upgrade**
     - C++17 → C++20 for concepts support and future backend dispatch improvements.

  7. **Bug fixes**
     - `exponential_`: pass `1.0 / lambd` (rate) instead of `lambd` to kernel.
     - `device_info.cpp`: conditional compilation to avoid CUDA API calls on non-CUDA backends.
     - `nonzero`: allocate one extra dummy row to prevent masked scatter lane corruption on NPU.
### Progress

- [ ] Change is properly reviewed (1 reviewer required, 2 recommended).
- [ ] Change is responded to an issue.
- [ ] Change is fully covered by a UT.

### CODE REVIEW

File: lib/exponential_.cpp
Comment (kiddyjinjin) [inline lib/exponential_.cpp:9]:
```suggestion
#include "flag_gems/backend_utils.h"
```
